internal/stategen/main.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// main is the entrypoint for the stategen tool, which updates a Librarian state.yaml
+// file to include the specified modules. The first argument is a path to the repository
+// root; all subsequent arguments are module names.
+func main() {
+	logLevel := slog.LevelInfo
+	if os.Getenv("GOOGLE_SDK_GO_LOGGING_LEVEL") == "debug" {
+		logLevel = slog.LevelDebug
+	}
+	slog.SetDefault(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+		Level: logLevel,
+	})))
+	slog.Info("stategen: invoked", "args", os.Args)
+	if err := run(os.Args[1:]); err != nil {
+		slog.Error("stategen: failed", "error", err)
+		os.Exit(1)
+	}
+	slog.Info("stategen: finished successfully")
+}
+
+func run(args []string) error {
+	if len(args) < 2 {
+		return errors.New("stategen: expected a root directory and at least one module")
+	}
+	repoRoot := args[0]
+	stateFilePath := filepath.Join(repoRoot, ".librarian/state.yaml")
+	state, err := parseLibrarianState(stateFilePath)
+	if err != nil {
+		return err
+	}
+	for _, moduleName := range args[1:] {
+		if stateContainsModule(state, moduleName) {
+			slog.Info("skipping existing module", "module", moduleName)
+			continue
+		}
+		err = addModule(repoRoot, state, moduleName)
+		if err != nil {
+			return err
+		}
+	}
+	return saveLibrarianState(stateFilePath, state)
+}
+
+func stateContainsModule(state *LibrarianState, moduleName string) bool {
+	for _, library := range state.Libraries {
+		if library.ID == moduleName {
+			return true
+		}
+	}
+	return false
+}
+
+func addModule(repoRoot string, state *LibrarianState, moduleName string) error {
+	slog.Info("adding module", "module", moduleName)
+	moduleRoot := filepath.Join(repoRoot, moduleName)
+
+	// Start off with the basics which need
+	library := &LibraryState{
```

Comment:
I think this is fine for an initial pass, but I wonder if we can make this smarter by reading some of the existing configuration. There are two good sources of information in google-cloud-go.
1. The owlbot config: https://github.com/googleapis/google-cloud-go/blob/main/.github/.OwlBot.yaml
2. Our side configuration: https://github.com/googleapis/google-cloud-go/blob/main/internal/postprocessor/config.yaml

Reply:
I was aware of the OwlBot config, but hadn't looked at the side configuration.

I was generally reckoning that using this as a starting point and then manually tweaking the state file for "unusual" APIs would be easier than making the tooling smart enough to handle everything. Looking at OwlBot, I figured it would be relatively few APIs that needed tweaking that way.

The side config list of input directories may be a better option than using the snippet metadata to figure out the API paths though - is that one of the things you were thinking about?